src/WebJobs.Script/Description/FunctionInvokerBase.cs

```diff
@@ -81,7 +85,99 @@ protected bool InitializeFileWatcherIfEnabled()
             return false;
         }
 
-        public abstract Task Invoke(object[] parameters);
+        public Task Invoke(object[] parameters)
+        {
+            FunctionStartedEvent startedEvent = null;
+            Task invokeTask = null;
+
+            CancellationToken timeoutOrShutdownToken = (CancellationToken)parameters.First(p => p?.GetType() == typeof(CancellationToken));
+            ExecutionContext functionExecutionContext = (ExecutionContext)parameters.First(p => p?.GetType() == typeof(ExecutionContext));
+            string invocationId = functionExecutionContext.InvocationId.ToString();
+
+            startedEvent = new FunctionStartedEvent(functionExecutionContext.InvocationId, Metadata);
+            _metrics.BeginEvent(startedEvent);
+
+            try
+            {
+                TraceWriter.Info($"Function started (Id={invocationId})");
+
+                invokeTask = InvokeInternal(parameters);
+                invokeTask.Wait(timeoutOrShutdownToken);
```

Comment:
Let's try to avoid blocking here. Should be able to use something like `await Task.WhenAny` passing the function task and `Task.Delay(-1, timeoutOrShutdownToken)` instead